src/common/hist_util.cu

```diff
@@ -98,379 +43,413 @@ __global__ void UnpackFeaturesK(float* __restrict__ fvalues,
-                                   fvalues_cur_.begin(),
-                                   weights_.begin(),
-                                   weights2_.begin(),
-                                   num_elements_.begin(),
-                                   thrust::maximum<bst_float>(),
-                                   gpu_batch_nrows_);
-    tmp_size = std::max(tmp_size, cur_tmp_size);
-    // size for filtering
-    cub::DeviceSelect::If(nullptr,
-                          cur_tmp_size,
-                          fvalues_.begin(),
-                          fvalues_cur_.begin(),
-                          num_elements_.begin(),
-                          gpu_batch_nrows_,
-                          IsNotNaN());
-    tmp_size = std::max(tmp_size, cur_tmp_size);
-
-    tmp_storage_.resize(tmp_size);
+    return true;
   }
-
-  void Sketch(const SparsePage& row_batch, const MetaInfo& info) {
-    // copy rows to the device
-    dh::safe_cuda(cudaSetDevice(device_));
-    const auto& offset_vec = row_batch.offset.HostVector();
-    row_ptrs_.resize(n_rows_ + 1);
-    thrust::copy(offset_vec.data(), offset_vec.data() + n_rows_ + 1, row_ptrs_.begin());
-    size_t gpu_nbatches = common::DivRoundUp(n_rows_, gpu_batch_nrows_);
-    for (size_t gpu_batch = 0; gpu_batch < gpu_nbatches; ++gpu_batch) {
-      SketchBatch(row_batch, info, gpu_batch);
+  __device__ bool operator()(const Entry& e) const {
+    if (common::CheckNAN(e.fvalue) || e.fvalue == missing) {
+      return false;
     }
+    return true;
   }
+};
 
-  void SketchBatch(const SparsePage& row_batch, const MetaInfo& info, size_t gpu_batch) {
-    // compute start and end indices
-    size_t batch_row_begin = gpu_batch * gpu_batch_nrows_;
-    size_t batch_row_end = std::min((gpu_batch + 1) * gpu_batch_nrows_,
-                                    static_cast<size_t>(n_rows_));
-    size_t batch_nrows = batch_row_end - batch_row_begin;
-
-    const auto& offset_vec = row_batch.offset.HostVector();
-    const auto& data_vec = row_batch.data.HostVector();
-
-    size_t n_entries = offset_vec[batch_row_end] - offset_vec[batch_row_begin];
-    // copy the batch to the GPU
-    dh::safe_cuda(cudaMemcpyAsync(entries_.data().get(),
-                                  data_vec.data() + offset_vec[batch_row_begin],
-                                  n_entries * sizeof(Entry),
-                                  cudaMemcpyDefault));
-    // copy the weights if necessary
-    if (has_weights_) {
-      const auto& weights_vec = info.weights_.HostVector();
-      dh::safe_cuda(cudaMemcpyAsync(weights_.data().get(),
-                                    weights_vec.data() + batch_row_begin,
-                                    batch_nrows * sizeof(bst_float),
-                                    cudaMemcpyDefault));
-    }
+template <typename ReturnT, typename IterT, typename FuncT>
+thrust::transform_iterator<FuncT, IterT, ReturnT> MakeTransformIterator(
```

Comment:
I think thrust has a similar function.

Reply:
I wrote this because of issues I used to have with msvc where it could not deduce the return type for the thrust version. I tried it without and it seems to be resolved now :)

Reply:
Actually I get a fatal error with thrust::make_transform_iterator so there seems to be problems still. I left this helper function in with a note.